Add File Field to Form (cleaned up)

Soare-Robert-Daniel opened the pull request:
<!-- Issues that this pull request closes. -->
Closes #1466.

This is the clean-up version of #1516 
<!-- Should look like this: `Closes #1, #2, #3.` . -->

### Summary
<!-- Please describe the changes you made. -->

Add File Field for Pro users.

### Screenshots <!-- if applicable -->

How it looks

![image](https://user-images.githubusercontent.com/17597852/225253957-0105a8ba-75a4-48a9-a2ac-deaea0f4ca98.png)

An email with an image saved to Media Library

![image](https://user-images.githubusercontent.com/17597852/225254490-8e3a1782-1ed7-48e5-92dc-5625ded7134c.png)

An email with the image as an attachment.

![image](https://user-images.githubusercontent.com/17597852/225254692-98231863-9e80-496b-b952-26d3fb0807cc.png)


----

### Test instructions
<!-- Describe how this pull request can be tested. -->

ℹ️ You will need to have a SMTP connection. You use can [WP SMPT plugin](https://wordpress.org/plugins/wp-mail-smtp/) + [GMAIL API](https://wpforms.com/how-to-securely-send-wordpress-emails-using-gmail-smtp/)

Example of testing for https://github.com/Codeinwp/otter-blocks/pull/1506


https://user-images.githubusercontent.com/17597852/228494364-5f9f00df-914c-4c5d-8cbf-74b6053eba4f.mp4

#### Form Field

ℹ️ This is a PRO feature, Free user should see an Upsell. Only Pro user can use it.

This feature has a lot of checks:
- File Size
- File Type
- Files Number

They are implemented in both frontend (you will see messages on the input field) and the backend (you will see them as detailed errors in the Console and a simple message at the bottom of the form).

Checking the errors from the backend is more complicated since you will have a special script to run from outside the page. [link](https://github.com/Codeinwp/otter-blocks/pull/1557#issuecomment-1469842489)

ℹ️ Images of type JPG/JPEG are both visible in the picked. If you want only JPG images, do not expect the file picker to see only JPG images, JPEG will also be visible, but they will be rejected since only JPG is allowed. This is a quirk of the browser.

Video example on saving images on media file, then switching to email attachment


https://user-images.githubusercontent.com/17597852/228497352-e97b146a-1f51-4591-bf16-b70d51350b16.mp4



<!--
#### Query
```javascript
new QueryQA().select('blocks').run()
```
-->

---- 

### Checklist before the final review

- [ ] Visual elements are not affected by independent changes.
- [ ] It is at least compatible with the [minimum WordPress version](https://wordpress.org/plugins/otter-blocks/).
- [ ] It loads additional script in frontend only if it is required.
- [ ] Does not impact the [Core Web Vitals](https://web.dev/vitals/).
- [ ] In case of deprecation, old blocks are safely migrated.
- [ ] It is usable in Widgets and FSE.
- [ ] Copy/Paste is working if the attributes are modified.

HardeepAsrani reviewed (changes requested):
As mentioned on Slack before, this feature doesn't work for me. Once I switch the Field type, I don't see any File Upload related controls.

<img width="1716" alt="Screenshot 2023-03-15 at 12 39 00 AM" src="https://user-images.githubusercontent.com/2649903/225111678-8d701e89-8b2d-4cfc-bd85-465df08175eb.png">

Soare-Robert-Daniel commented:
# Testing

### Testing script (Run in browser)

This script test if the backend gives the correct errors based on the payload. If you set the limit to 10 files, but this script sends 20 files, and the request is successful, then it is a problem ( it should have give error code `14`) 

```javascript
( ()=>{

	// +------------------ Start Settings ------------------+

	// Modify those values based on your needs
	const formOptionName = 'bf9b28ff82ce323187abc5d0feb0d8f07ecd8436_d3188243'; // This is an example. Change it with the value of form that you want to target

	// The general structure of the data.
	const fileFieldMockData =
    {
    	label: 'Test files',
    	value: 'test.png image',
    	type: 'file',
    	metadata: {
    		name: 'test.png',
    		size: 1000,
    		data: '',
    		fieldOptionName: null
    	}
    };

	const invalidFileExtensionTest = 'image/aros'; // Set the extension that you want test.
	const filesNumberTest = 20; // Set the number of files that you want to test.
	const fileSizeTest = 1024 * 1024 * 10; // Set the aproximate size for the file that you want to test. Random number. A message will apear with the size generated.

	// +------------------ End Settings ------------------+

	// Select the form
	const form = document.querySelector( `div[data-option-name=${formOptionName}]` );

	if ( ! form ) {
		console.warn( `Form with option name ${formOptionName} not found. You might have a type or it does not exists in page!` );
		return;
	}

	console.log( 'Found Form:', form );

	// Get the first File field
	const fileField = form.querySelector( 'input[type="file"]' );

	if ( ! fileField ) {
		console.warn( 'There is no File field in form! Add a File field to it.' );
		return;
	}

	const fieldOptionName = fileField?.dataset?.fieldOptionName;

	if ( ! fieldOptionName ) {
		console.warn( 'This field is not registered.' );
		return;
	}

	// Use the File found. Without a this the request will be rejected.
	fileFieldMockData.metadata.fieldOptionName = fieldOptionName;

	console.log( `Found File field with option name ${fieldOptionName}`, fileField );

	// Prepare payload for the Too Many File Test
	const multipleFilesPayload = Array( 20 ).fill( fileFieldMockData );

	// Prepare payload for File Too Big Test.
	console.log( 'Generating big file...' );
	const bigFilePayload = [ fileFieldMockData ].map( p => {
		const data = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIAQMAAAD+wSzIAAAABlBMVEX///+/v7+jQ3Y5AAAADklEQVQI12P4AIX8EAgALgAD/aNpbtEAAAAASUVORK5CYII' + btoa( ( Array( fileSizeTest ).fill( 'f' ) ) ); // a valid encoded image + a very long string
		const size = new Blob([ data ]).size;
		console.log( `Created a file with the size of ${ ( size / ( 1024 * 1024 ) ).toFixed( 2 )}MB` );
		return { ...p, metadata: { ...p.metadata, data: data, size: size }};
	});

	// Prepare the payload for Invalid File Type Test.
	const invalidFileTypePayload = [ fileFieldMockData ].map( p => {
		return { ...p, metadata: { ...p.metadata, data: `data:${invalidFileExtensionTest};base64,adfadsfdsuihdashfkdjasfhlkashf` }};
	});

	console.log( 'Payloads', { multipleFilesPayload, bigFilePayload, invalidFileTypePayload });

	const lunchRequest = ( formFieldsData, message ) => {
		const payload = {};
		payload.formInputsData = formFieldsData;
		if ( '' !== form?.dataset?.emailSubject ) {
			payload.emailSubject = form?.dataset?.emailSubject;
		}

		if ( form?.dataset?.optionName ) {
			payload.formOption = form?.dataset?.optionName;
		}

		if ( form?.id ) {
			payload.formId = form?.id;
		}

		payload.nonceValue = form.querySelector( `.protection #${form.id || ''}_nonce_field` )?.value;


		payload.antiSpamTime = Date.now();
		payload.antiSpamHoneyPot = Boolean( form.querySelector( ':scope > .otter-form__container > .protection .o-anti-bot' )?.checked ?? false );

		payload.postUrl = window.location.href;


		/**
	 * Get the consent
	 */
		if ( form.classList.contains( 'can-submit-and-subscribe' ) ) {
			payload.action = 'submit-subscribe';
			payload.consent = form.querySelector( '.otter-form-consent input' )?.checked || false;
		}

		const formURlEndpoint = ( window?.themeisleGutenbergForm?.root || ( window.location.origin + '/wp-json/' ) ) + 'otter/v1/form/frontend';

		console.log( `Send payload for test: "${message}"`, { payload });

		fetch( formURlEndpoint, {
			method: 'POST',
			headers: {
				'Content-Type': 'application/json',
				Accept: 'application/json, */*;q=0.1',
				'X-WP-Nonce': window?.themeisleGutenbergForm?.nonce
			},
			credentials: 'include',
			body: JSON.stringify({
				handler: 'submit',
				payload
			})
		})
			.then( r => r.json() )
			.then( r => {
   	console.log( message, {
					result: r,
					errorCode: r?.code,
					reasons: r?.reasons
				});
			});
	};

	lunchRequest( multipleFilesPayload, 'Multiple files test' );
	lunchRequest( invalidFileTypePayload, 'Invalid file test' );
	lunchRequest( bigFilePayload, 'File Size to big test' );

})();
```

### Usage

How the run should look.

<img width="1055" alt="Screenshot 2023-03-15 at 12 54 34" src="https://user-images.githubusercontent.com/17597852/225288400-50134461-2851-4ce3-8c2a-eaf0a68bb03b.png">

Video

https://user-images.githubusercontent.com/17597852/225296719-d0f94f6e-d5c9-48b7-904e-d0e04ba879f3.mp4

HardeepAsrani commented on inc/integrations/api/form-response-data.php:
```diff
@@ -280,77 +289,118 @@ public function set_is_credential_error( $is_credential_error ) {
+		switch ( $error_code ) {
 			case self::ERROR_MISSING_DATA:
-				$this->add_reason( __( 'Essential data is missing: invalid Form id or protection.', 'otter-blocks' ) );
+				$messages[] = ( __( 'Essential data is missing: invalid Form id or protection.', 'otter-blocks' ) );
```
We can remove the extra braces, don't really need.

HardeepAsrani commented on inc/integrations/api/form-response-data.php:
```diff
@@ -280,77 +289,118 @@ public function set_is_credential_error( $is_credential_error ) {
+	 * @since 2.2.3
+	 */
+	public static function get_error_code_message( $error_code ) {
+		$messages = array();
```
Why is this an array? I only see $messages returning a single string. Is there a reason?

Also, you can use an array instead of a switch like this:

```
$error_messages = array(
	self::ERROR_MISSING_DATA => __( 'Essential data is missing: invalid Form id or protection.', 'otter-blocks' ),
	self::ERROR_MISSING_CAPTCHA => __( 'Captcha token is missing.', 'otter-blocks' ),
	// And so on
);

if ( ! isset( $error_messages[ $error_code ] ) ) {
	return "Expected error whatever message";
}

return $error_messages[ $error_code ];
```

Soare-Robert-Daniel replied:
The main idea was to allow multiple error composition (not the case at the current complexity).

If we do more integration, a lot of errors will overlap. So instead of making a separate string for every error, we can compose like: `array( 'Provider Mailchimp', 'Invalid API Key')` or `array( 'Provider Sendinblue', 'Invalid API Key')`

Or course, we can concatenate the string and return a single one, but I liked the idea of an array, especially if we will create some more complex error handling.

HardeepAsrani replied:
Until we do that we can just keep it simple. Not the most important thing.

HardeepAsrani commented on inc/plugins/class-dynamic-content.php:
Once again this PR is reverting Arina's PR: https://github.com/Codeinwp/otter-blocks/commit/ba635975b4a0ad710f8dc8d7c8e60e563e490673

HardeepAsrani commented on inc/render/class-form-file.php:
I'd keep this in the Pro so we can simply edit the file and have the new changes even if at times Otter free is older version than the Pro.

HardeepAsrani commented on inc/render/class-review-block.php:
This is once again, undoing the changes Arina made in a separate PR.

HardeepAsrani commented on src/blocks/blocks/form/file/block.json:
We can also keep this to Pro. The logic here being if there is a bug that we fix in Form, we can be assured it will work even if free version is one version behind, so it's good keeping all the Form File stuff in Pro.

HardeepAsrani commented on src/css/editor.js:
This is undoing Arina's changes again. This is a dangerous pattern, specially if we miss some small changes in the process which can break things. We're creating regressions that we don't need.

HardeepAsrani commented on src/pro/plugins/form/index.js:
```diff
@@ -0,0 +1,125 @@
+				value={ attributes.allowedFileTypes }
+				onChange={ allowedFileTypes => setAttributes({ allowedFileTypes }) }
+				help={ __( 'Add the allowed files types that can be loaded. E.g.: .png, .mp4, .jpg, .zip, .pdf. Attention: The host provider might not allow to saving of all type of files.', 'otter-blocks' ) }
+				suggestions={ [
```
We can put it in a constant for better readability.

HardeepAsrani reviewed (changes requested):
I've left feedback on files. Apart from that:

- If you're using Pro, you don't need to have (Pro) label in the Select input.
- If you're using free, there's no link to upsell, also the Label fields seem to be editable. This makes it look bit weird and feels like the field is still usable just with limitations. I'd avoid showing the field preview in the editor, just show the sidebar having locked options with an Upsell.

Also the biggest issue I see is us reverting older commits, I've left comments on the ones I identified but it should be made to that nothing is being reverted.

HardeepAsrani commented on src/blocks/blocks/font-awesome-icons/style.scss:
```diff
@@ -11,7 +11,6 @@
 
 	display: flex;
 	justify-content: var(--align);
-	margin: 0;
```
This is still undoing many commits.

HardeepAsrani commented on src/blocks/blocks/button-group/style.scss:
```diff
@@ -113,7 +113,7 @@ $device-map: (desktop: '1200px', tablet: '960px', mobile: '600px');
 	--padding: 15px 30px;
 	--padding-tablet: var(--padding);
 	--padding-mobile: var(--padding-tablet);
-	--font-size: var( --btnfs, var(--bodyfontsize, 15px) );
```
Same here.

HardeepAsrani commented on src/blocks/blocks/accordion/style.scss:
```diff
@@ -108,7 +108,7 @@
 	// when no gap is set, we need to avoid the doubled border between items
 	&:not( .is-style-default ).no-gap > &-item:not( :first-child ),
 	&:not( .is-style-default ):not( .is-narrow-gap ):not( .is-wide-gap ):not( .is-wider-gap ):not( .has-gap ) > &-item:not( :first-child ) {
-		&> #{$block}-item__title {
```
Here as well.

HardeepAsrani reviewed (changes requested):
It is still undoing some older commits.

irinelenache commented:
@Soare-Robert-Daniel I tried to test this PR but immediately after activating Otter pro, i'm getting a fatal error https://vertis.d.pr/i/H6lur5
```
Warning:  require(/s4-otter226/wordpress/wp-content/plugins/otter-pro/inc/render/class-form-file-block.php): failed to open stream: No such file or directory in /s4-otter226/wordpress/wp-content/plugins/otter-pro/autoloader.php on line 148
Fatal error:  require(): Failed opening required '/s4-otter226/wordpress/wp-content/plugins/otter-pro/inc/render/class-form-file-block.php' (include_path='.:/usr/share/php') in /s4-otter226/wordpress/wp-content/plugins/otter-pro/autoloader.php on line 148
```

irinelenache commented:
@Soare-Robert-Daniel Is the same, but now it won't let me activate the plugin https://vertis.d.pr/i/cRASMe

Soare-Robert-Daniel commented:
All working now. It seams that the loader expected a file name format.

irinelenache commented:
@Soare-Robert-Daniel Nice job 🚀 Managed to test and here is what i found:

- [ ] Adding a longer option in the select field will break change the column width https://vertis.d.pr/v/Z4PWSv
- [ ] The file type appears with Pro label even if i'm already using Otter pro. When i select it, the label disappears https://vertis.d.pr/v/NGlvI0
- [ ] The checkbox and radio inherit the size from the Label size https://vertis.d.pr/i/iziyyP
- [ ] The **Back to form** button doesn't work when you edit the file inputs
- [ ] If multiple files are allowed, they will appear as multiple fields https://vertis.d.pr/i/jURFRb , i think they should appear inline
- [ ] When some file is added to the email, the fields order will be broken https://vertis.d.pr/i/igHOsH
- [ ] If i'm using two file fields with different saving settings (one should be saved in media library, one should be sent on email), both of them will be saved in library
- [ ] `image/*` type works fine for me, but `.jpg` , `.jpeg` , `.jpg, .jpeg` will throw an error that file type is not allowed https://vertis.d.pr/i/BRvlmc

Soare-Robert-Daniel commented:
> * [x]  `image/*` type works fine for me, but `.jpg` , `.jpeg` , `.jpg, .jpeg` will throw an error that file type is not allowed https://vertis.d.pr/i/BRvlmc

I simplified this fiasco. The correct extensions name is `image/jpeg`. If you try to type `.jpg` or `image/jpg`, it will be replaced with `image/jpeg`. [More info about this](https://www.keycdn.com/support/difference-between-jpg-and-jpeg)

ℹ️ If you reuse an instance, change the previously used File inputs that have `.jpg` as extensions. If you modified the field, you will the renaming.

Video to show the switch

https://user-images.githubusercontent.com/17597852/230076670-84760578-64d4-4a5a-ba78-02faffef0f6e.mp4

Also, I added a new error when the File Input field options in the WordPress database are not saved correctly (or they may be deleted)

https://user-images.githubusercontent.com/17597852/230076908-9408919a-52c8-4b38-9d89-7f21e12649a4.mp4

Now multiple types of Field input work correctly in the same form. Also, the files in the email are now grouped together, and all the fields are now in order.

https://user-images.githubusercontent.com/17597852/230077408-b8765fe1-0c59-4a31-89af-11b6ef3bba81.mp4

HardeepAsrani reviewed (changes requested):
As discussed on Slack, isChanged attribute can be made into a state/synced with global variable instead of adding it into block attributes.

irinelenache commented:
@Soare-Robert-Daniel The mentioned issues are fixed now, thank you 🚀

Here are some other things that i found:

- [ ] For the **select** input, the **border width and radius** aren't applied https://vertis.d.pr/i/enCm2H I also saw that the **input field size** isn't applied
- [ ] For the file input, which file types are allowed exactly? I had `.csv, .pdf, .xlsx` set for a [field](https://vertis.d.pr/i/k0TLTI) and for every type i got this console error `(12) The file type is not allowed.`

Soare-Robert-Daniel commented:
> * [ ]  For the file input, which file types are allowed exactly? I had `.csv, .pdf, .xlsx` set for a [field](https://vertis.d.pr/i/k0TLTI) and for every type i got this console error `(12) The file type is not allowed.`

Thanks for pointing it out. The WP MIME checked did not like the `.`, I made a little cleanup on it.

Regarding the allowed file: all files to meet the condition specified in Form and _server_ are permitted. The server part is important, to save a file on the server, it must be allowed. It is the user's job to ensure the server can save it, something like this https://www.wpbeginner.com/wp-tutorials/how-to-add-additional-file-types-to-be-uploaded-in-wordpress/. Sometimes the hosting platform is banning files like `.exe`, then the submitted files are rejected.

Also, a more helpful error is now displayed:
![image](https://user-images.githubusercontent.com/17597852/230905637-b39f621c-5ad4-45f6-8a07-1fc90ae2f0ff.png)

For Code Errors:
- 12: Does not meet the condition from Form settings
- 13: Does not meet the condition from server.

---

@HardeepAsrani, do you think that File, Checkbox, Radio should have a background color, border, and padding like Text?

![image](https://user-images.githubusercontent.com/17597852/230906855-709b109b-7c8a-4846-8cd3-5e0f05a90c04.png)

HardeepAsrani commented:
> @HardeepAsrani, do you think that File, Checkbox, Radio should have a background color, border, and padding like Text?

No, they aren't needed.

Soare-Robert-Daniel commented:
@irinelenache, the issue about file format like `.zip` should be fixed.

Also, the File field will have the same level of styling as Checkbox, and Radio: no background color, no border, and no padding (can be seen in the video)

Video

https://user-images.githubusercontent.com/17597852/231114073-bfe973b3-a0c5-467c-a317-0d2d62663b1a.mp4

rodica-andronache commented:
@Soare-Robert-Daniel here's what I've seen while testing this PR. 

- [ ] https://www.loom.com/share/49f61652b62d40788350fe697e9a6ae3 - Invalid field for the URL field
- [ ] Save to media library gets reset all the time https://vertis.d.pr/v/SOdoN4
- [ ] https://www.loom.com/share/e3f7d005ee024ae0a6d1027f59484935 - issue with the File size
- [ ] https://vertis.d.pr/i/MawYlE - Is not very intuitive how to add options in the editor ( when you click enter, it doesn’t add a new option) - not sure if there's something to do about it.

P.S [This other PR](https://github.com/Codeinwp/otter-blocks/pull/1506) has some extra issues. I shouldn't worry about those, right? Thanks!

rodica-andronache commented:
@Soare-Robert-Daniel sure, we can test again after the PR is merged, as there are indeed lots of stuff that break 👍 

For now, on this PR, the only remaining thing I've seen is this one:
- For the **Form** block, go to the **Style** tab > **Input Labels** ( but same thing happens for the other options as well ) > **Fields Spacing**. There are some layout issues. I'm testing on a 14-inch MacBook Pro
This is how it looks with this PR https://vertis.d.pr/i/nJ1oj2
This is how it looks with the current version of Otter https://vertis.d.pr/i/KwUXLJ

Apart from that, a small question. https://vertis.d.pr/i/ZzNO90 - is it expected for users to be able to add anything in the **Allowed File Types** field?
--------

For https://github.com/Codeinwp/otter-blocks/pull/1506, I can't download the build anymore https://vertis.d.pr/i/7MhvGF
but this is the issue I saw the last time I checked the PR.

https://vertis.d.pr/v/4IBvHA
If I go **Back to Form Fields**, all fields appear as **Textarea**, you **can’t edit them or delete them**
